examples/ShoppingAds/AddPerformanceMaxProductListingGroupTree.php

```diff
@@ -0,0 +1,602 @@
+            if (array_key_exists($parentResourceName, $parentsToChildren)) {
+                $siblings = $parentsToChildren[$parentResourceName];
+            }
+            $siblings[] = $assetGroupListingGroupFilter->getResourceName();
+            $parentsToChildren[$parentResourceName] = $siblings;
+        }
+
+        // [START add_performance_max_product_listing_group_tree_2]
+        return self::createMutateOperationsForRemovingDescendents(
+            $rootResourceName,
+            $parentsToChildren
+        );
+        // [END add_performance_max_product_listing_group_tree_2]
+    }
+
+    /**
+     * Creates a list of mutate operations that remove all the descendents of the specified
+     * asset group listing group filter's resource name. The order of removal is post-order,
+     * where all the children (and their children, recursively) are removed first. Then,
+     * the node itself is removed.
+     *
+     * @param string $assetGroupListingGroupFilterResourceName the resource name of the root of
+     *     listing group tree
+     * @param array $parentsToChildren the map from parent resource names to children resource
+     *     names
+     * @return MutateOperation[] the list of MutateOperations to remove all listing groups
+     */
+    // [START add_performance_max_product_listing_group_tree_3]
+    private static function createMutateOperationsForRemovingDescendents(
+        string $assetGroupListingGroupFilterResourceName,
+        array $parentsToChildren
+    ): array {
+        $operations = [];
+        if (array_key_exists($assetGroupListingGroupFilterResourceName, $parentsToChildren)) {
+            foreach ($parentsToChildren[$assetGroupListingGroupFilterResourceName] as $child) {
+                $operations = array_merge(
+                    $operations,
+                    self::createMutateOperationsForRemovingDescendents($child, $parentsToChildren)
+                );
+            }
+        }
+
+        $operations[] = new MutateOperation([
+            'asset_group_listing_group_filter_operation'
+                => new AssetGroupListingGroupFilterOperation([
+                    'remove' => $assetGroupListingGroupFilterResourceName
+                ])
+        ]);
+        return $operations;
+    }
+    // [END add_performance_max_product_listing_group_tree_3]
+
+    /**
+     * Creates a mutate operation that creates a root asset group listing group filter for the
+     * factory's asset group.
+     *
+     * The root node or partition is the default, which is displayed as "All Products".
+     *
+     * @param int $customerId the customer ID
+     * @param int $assetGroupId the asset group ID
+     * @param int $rootListingGroupId the root listing group ID
+     * @return MutateOperation the mutate operation for creating the root
+     */
+    // [START add_performance_max_product_listing_group_tree_4]
+    private static function createMutateOperationForRoot(
```

Comment:
This function seems to be unnecessary, should we refactor to remove it altogether?

If we want to keep it, I suggest duplicating the content of `createMutateOperationForSubdivision` and removing the support of the special case for root in `createMutateOperationForSubdivision` for easier understanding.